sgx/common/src/retrieve_key_protocol.c

```diff
@@ -0,0 +1,1164 @@
+/**
+ * @file retrieve_key_protocol.c
+ *
+ * @brief Provides implementation for functionality directly supporting
+ *        creating, validating, and parsing the kmyth 'retrieve key'
+ *        protocol messages.
+ */
+
+#include "retrieve_key_protocol.h"
+
+/*****************************************************************************
+ * extract_identity_bytes_from_x509()
+ ****************************************************************************/
+int extract_identity_bytes_from_x509(X509 *cert_in, X509_NAME **identity_out)
+{
+  // extract 'subject name' from input certificate
+  //   Note: The returned X509_NAME is an internal pointer
+  //         that should NOT be freed.
+  X509_NAME *subj_name = X509_get_subject_name(cert_in);
+  if (subj_name == NULL)
+  {
+    kmyth_sgx_log(LOG_ERR, "extraction of certificate's subject name failed");
+    return EXIT_FAILURE;
+  }
+
+  *identity_out = X509_NAME_dup(subj_name);
+  if (identity_out == NULL)
+  {
+    kmyth_sgx_log(LOG_ERR, "duplication of X509_NAME struct failed");
+    return EXIT_FAILURE;
+  }
+
+  return EXIT_SUCCESS;
+}
+
+/*****************************************************************************
+ * append_msg_signature()
+ ****************************************************************************/
+int append_msg_signature(EVP_PKEY * sign_key,
+                         ECDHMessage * msg)
+{
+  // compute message signature
+  unsigned char *signature_bytes = NULL;
+  int signature_len = 0;
+
+  if (EXIT_SUCCESS != ec_sign_buffer(sign_key,
+                                     msg->body,
+                                     msg->hdr.msg_size,
+                                     &signature_bytes,
+                                     &signature_len))
+  {
+    kmyth_sgx_log(LOG_ERR, "error signing buffer");
+    free(signature_bytes);
+    return EXIT_FAILURE;
+  }
+
+  // create a temporary copy of the input message
+  size_t buf_copy_len = msg->hdr.msg_size;
+  unsigned char *buf_copy = malloc(buf_copy_len);
```

Comment:
In Kmyth, we normally check the output of malloc and related functions to make sure it isn't NULL, but that doesn't happen here and in several other places in `retrieve_key_protocol.c`.

Reply:
Attempted to fix this in commit [22e49d4](https://github.com/NationalSecurityAgency/kmyth/pull/168/commits/22e49d407672290bbf22260defa877af0126d31f).